synapse/storage/databases/main/events_worker.py

```diff
@@ -1626,3 +1626,189 @@ def _cleanup_old_transaction_ids_txn(txn):
+            # prev_events yet. We need to check whether those prev_events are
+            # backward extremity also known as gaps that need to be backfilled.
+            backward_extremity_query = """
+                SELECT 1 FROM event_backward_extremities
+                WHERE
+                    room_id = ?
+                    AND event_id = ?
+                LIMIT 1
+            """
+
+            # If the event in question has its prev_events listed as a backward
+            # extremity, it's next to a backwards gap.
+            for prev_event_id in event.prev_event_ids():
+                txn.execute(backward_extremity_query, (event.room_id, prev_event_id))
+                backward_extremities = txn.fetchall()
+
+                # We consider any backward extremity as a backwards gap
+                if len(backward_extremities):
+                    return True
+
+            return False
+
+        return await self.db_pool.runInteraction(
+            "is_event_next_to_backward_gap_txn",
+            is_event_next_to_backward_gap_txn,
+        )
+
+    async def is_event_next_to_forward_gap(self, event: EventBase) -> bool:
+        """Check if the given event is next to a forward gap of missing events.
+        The gap in front of the latest events is not considered a gap.
+        <latest messages> A(False)--->B(False)--->C(False)--->  <gap, unknown events> <oldest messages>
+        <latest messages> A(False)--->B(False)--->  <gap, unknown events>  --->D(True)--->E(False) <oldest messages>
+
+        Args:
+            room_id: room where the event lives
+            event_id: event to check
+
+        Returns:
+            Boolean indicating whether it's an extremity
+        """
+
+        def is_event_next_to_gap_txn(txn) -> bool:
+            # If the event in question is a forward extremity, we will just
+            # consider any potential forward gap as not a gap since it's one of
+            # the latest events in the room.
+            #
+            # We can't combine this query with the `forward_edge_query` below
+            # because if the event in question has no forward edges(isn't
+            # referenced by any other event's prev_events) but is in
+            # `event_forward_extremities`, we don't want to return 0 rows and
+            # say it's next to a gap.
+            forward_extremity_query = """
+                SELECT 1 FROM event_forward_extremities
+                WHERE
+                    room_id = ?
+                    AND event_id = ?
+                LIMIT 1
+            """
+
+            # Check to see whether the event in question is already referenced
+            # by another event. If we don't see any edges, we're next to a
+            # forward gap.
+            #
+            # We can't just only check `event_forward_extremities` directly
+            # because that doesn't include backfilled and outlier events.
```

Comment:
This comment is confusing. It seems to imply that `event_forward_extremities` almost does what we want, when it's the opposite - events in that list are considered to not be next to gaps.

```suggestion
```

Reply:
Good point, it's a bit off. I've added the clarification to the `forward_extremity_query` comments instead and re-framed it to explain why we can't just use it by itself.

```
# `event_forward_extremities` does not include backfilled or outlier
# events so we can't rely on it to find forward gaps. We can only
# use it to determine whether a message is the latest in the room.
```